examples/fungible-token/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        Ok(st.symbol().to_string())
+    }
+
+    fn total_supply<BS, RT>(rt: &mut RT) -> Result<TokenAmount, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        Ok(st.total_supply().clone())
+    }
+
+    fn balance<BS, RT>(rt: &mut RT, addr: Address) -> Result<TokenAmount, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let id = rt
+            .resolve_address(&addr)
+            .ok_or_else(|| actor_error!(illegal_argument; "cannot resolve account address"))?
+            .id()
+            .unwrap();
+        let st: State = rt.state()?;
+        let b = st.balance(rt.store(), id).map_err(|e| {
+            e.downcast_default(
+                ExitCode::USR_ILLEGAL_STATE,
+                "failed to get balance from store",
+            )
+        })?;
+        Ok(b)
+    }
+
+    fn transfer<BS, RT>(rt: &mut RT, params: TransferParams) -> Result<TransferReturn, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let TransferParams { to, amount } = params;
+
+        let from_id = rt.message().caller().id().unwrap();
+        let to_id = rt
+            .resolve_address(&to)
+            .ok_or_else(
+                || actor_error!(illegal_argument; "cannot resolve destination account address"),
+            )?
+            .id()
+            .unwrap();
+
+        let (from_balance, to_balance) = rt.transaction(|st: &mut State, rt| {
+            st.transfer(rt.store(), from_id, to_id, amount)
+                .map_err(|e| {
+                    e.downcast_default(ExitCode::USR_UNSPECIFIED, "cannot perform transfer")
+                })
+        })?;
+
+        Ok(TransferReturn {
+            from_balance,
+            to_balance,
+        })
+    }
+
+    fn init_atomic_transfer<BS, RT>(
+        rt: &mut RT,
+        params: InitAtomicTransferParams,
+    ) -> Result<AtomicInputID, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let InitAtomicTransferParams {
+            coordinator,
+            transfer: TransferParams { to, amount },
+        } = params;
+
+        let from_id = rt.message().caller().id().unwrap();
+        let to_id = rt
+            .resolve_address(&to)
+            .ok_or_else(
+                || actor_error!(illegal_argument; "cannot resolve destination account address"),
+            )?
+            .id()
+            .unwrap();
+
+        let input_id = rt.transaction(|st: &mut State, rt| {
+            st.init_atomic_transfer(rt.store(), coordinator, from_id, to_id, amount)
+                .map_err(|e| {
+                    e.downcast_default(ExitCode::USR_UNSPECIFIED, "cannot init atomic transfer")
+                })
+        })?;
+        Ok(input_id)
+    }
+
+    fn cancel_atomic_transfer<BS, RT>(
```

Comment:
Cancel can be triggered by a user after initiating and before preparing. After prepare, the only way to cancel an execution is by calling the coordinator to trigger a rollback, right?

Reply:
Almost right, one can abort an already prepared atomic execution by invoking the `AbortAtomicTransfer` method of this actor, not the coordinator directly.

Reply:
But the abortion would have to propagated to the coordinator to inform it about this, right? (again, the protocol has a lot of moving parts, it would be great to have some comments to support the code :pray: )

Reply:
There are only two callable methods in the coordinator actor: `PreCommit` and `Revoke`. When cancelling an atomic execution that has not yet been pre-committed by the executing actor, as opposed to aborting an already pre-committed one, there is nothing to revoke at the coordinator actor, so the executing actor just consumes its input ID and releases the lock. After that, the input ID will not be accepted by the actor for pre-commitment anymore.